### PR TITLE
Feature: Add --no-cache flag to astro deploy and astro dev start

### PR DESCRIPTION
**Pull request by AmFlint**
## Description

- Add `--no-cache` flag on `astro dev start` and `astro deploy` commands. This allows users to force building docker images from scratch (ignore docker build cache).
- Had to refactor the ContainerHandler and ImageHandler interfaces and implementations for Docker and Podman inside the `airflow` package in order to accept the NoCache parameter.

## 🎟 Issue(s)

https://github.com/astronomer/issues/issues/3203

## 🧪 Functional Testing

**To test the --no-cache on astro deploy**:
- Create an Astronomer workspace, and `astro workspace switch` to this namespace
- Create an Astronomer deployment
- Use the astro deploy with --no-cache flag, you'll have to deploy at least 2 times, to check that the second time, the image is built again without cache:
```bash
astro deploy <DEPLOYMENT> --no-cache
```

**To test the --no-cache on astro dev start**:
- Initialize a local project with `astro init`
- Start the local project once (this will build the image without cache as this is the first build):
```bash
astro dev start
```
- Start a second time to check that the command uses cache (logs should print `CACHED` when building the image):
```bash
astro dev start
```
- Start a third time with the --no-cache flag (this should build the image from scratch without using docker build cache):
```bash
astro dev start
```

## 📋 Checklist

- [x] Rebased from the main (or release if patching) branch (before testing)
- [x] Ran `make test` before taking out of draft
- [x] Added/updated applicable tests
- [x] Tested against [Houston-API](https://github.com/astronomer/houston-api/) and [Astronomer](https://github.com/astronomer/astronomer/) (if necessary).
- [ ] Communicated to/tagged owners of respective clients potentially impacted by these changes.
- [ ] Updated any related [documentation](https://github.com/astronomer/docs/)

### CODE REVIEW

File: airflow/docker.go [hunk 1]
```diff
@@ -12,6 +12,8 @@ import (
 	"text/tabwriter"
 	"time"
 
+	container_types "github.com/astronomer/astro-cli/airflow/types"
```
Comment (neel-astro) [inline airflow/docker.go:15]: nitpick: `container_types` => `containerTypes`

File: airflow/docker_image.go [hunk 2]
```diff
@@ -10,6 +10,8 @@ import (
 	"os"
 	"os/exec"
 
+	container_types "github.com/astronomer/astro-cli/airflow/types"
```
Comment (neel-astro) [inline airflow/docker_image.go:13]: same here, `container_types` => `containerTypes`

File: cmd/airflow.go [hunk 3]
```diff
@@ -146,6 +149,7 @@ func newAirflowStartCmd(out io.Writer) *cobra.Command {
 		RunE:    airflowStart,
 	}
 	cmd.Flags().StringVarP(&envFile, "env", "e", ".env", "Location of file containing environment variables")
+	cmd.Flags().BoolVarP(&ignoreCacheDev, "no-cache", "", false, "Do not use cache when building docker image")
```
Comment (neel-astro) [inline cmd/airflow.go:152]: nitpick: `docker` => `container`

File: cmd/deploy.go [hunk 4]
```diff
@@ -61,6 +65,7 @@ func newDeployCmd() *cobra.Command {
 	cmd.Flags().BoolVarP(&forceDeploy, "force", "f", false, "Force deploy if uncommitted changes")
 	cmd.Flags().BoolVarP(&forcePrompt, "prompt", "p", false, "Force prompt to choose target deployment")
 	cmd.Flags().BoolVarP(&saveDeployConfig, "save", "s", false, "Save deployment in config for future deploys")
+	cmd.Flags().BoolVarP(&ignoreCacheDeploy, "no-cache", "", false, "Do not use cache when building docker image")
```
Comment (neel-astro) [inline cmd/deploy.go:68]: same here: `docker` => `container`

**Review by neel-astro — APPROVED** [review]
lgtm 👍, have added a few nitpicking comments do have a look

**Comment by snolan-amount** [comment]
I'd like to suggest also adding the docker build --pull option - it is related, sometimes you really need to pull the base image again, not just ignore the locally cached layers.

**Pull request merged**